### PR TITLE
Feature/5819 improvements caching

### PR DESCRIPTION
**Pull request by skamphuis**
This PR relats to #5819 and covers the caching-related changes:
- Fix serialization errors when using external cache
- HTML/Text: implement caching for GetWorkflows
- Only do a clearcache when unknown tab belongs to the portal whose cache is being flushed.
- TabController: only clear cache when tab is for the current portal

### CODE REVIEW

**Review by mitchelsellers — APPROVED** [review]
This looks good to me!

File: DNN Platform/Modules/HTML/Components/WorkflowStateController.cs [hunk 1]
```diff
@@ -13,16 +13,31 @@ namespace DotNetNuke.Modules.Html
     public class WorkflowStateController
     {
         private const string WORKFLOWCACHEKEY = "Workflow{0}";
+        private const string WORKFLOWS_CACHE_KEY = "Workflows{0}";
```
Comment (valadas) [inline DNN Platform/Modules/HTML/Components/WorkflowStateController.cs:16]: Our current stylecop rules prevent us from using _ (underscores) for these variables which is making the build fail currently.

File: DNN Platform/Modules/HTML/Components/WorkflowStateController.cs [hunk 2]
```diff
@@ -13,16 +13,31 @@ namespace DotNetNuke.Modules.Html
...
+        }
+
+        /// -----------------------------------------------------------------------------
+        /// <summary>
+        ///   GetWorkflowsCallBack retrieves a collection of WorkflowStateInfo objects for the Portal from the database.
+        /// </summary>
+        /// <param name = "cacheItemArgs">Arguments passed by the GetWorkflowStates method.</param>
+        /// <returns>WorkflowStateInfo List.</returns>
+        /// -----------------------------------------------------------------------------
+        public object GetWorkflowsCallBack(CacheItemArgs cacheItemArgs)
```
Comment (valadas) [inline DNN Platform/Modules/HTML/Components/WorkflowStateController.cs:37]: Do we _have_ to return an object here ?

**Review by valadas — CHANGES_REQUESTED** [review]

File: DNN Platform/Modules/HTML/Components/WorkflowStateController.cs (hunk 1 above)
Comment (skamphuis) [reply DNN Platform/Modules/HTML/Components/WorkflowStateController.cs:16]: Right, how did I not notice that...

File: DNN Platform/Modules/HTML/Components/WorkflowStateController.cs [hunk 3]
```diff
@@ -13,16 +13,31 @@ namespace DotNetNuke.Modules.Html
...
         private const CacheItemPriority WORKFLOWCACHEPRIORITY = CacheItemPriority.Normal;
 
-        /// <summary>GetWorkFlows retrieves a collection of workflows for the portal.</summary>
+        /// <summary>GetWorkFlows retrieves a collection of workflows for the portal from the cache.</summary>
         /// <param name="portalID">The ID of the Portal.</param>
         /// <returns>An <see cref="ArrayList"/> of <see cref="WorkflowStateInfo"/> instances.</returns>
         public ArrayList GetWorkflows(int portalID)
         {
-            return CBO.FillCollection(DataProvider.Instance().GetWorkflows(portalID), typeof(WorkflowStateInfo));
+            var cacheKey = string.Format(WORKFLOWSCACHEKEY, portalID);
```
Comment (bdukes) [inline DNN Platform/Modules/HTML/Components/WorkflowStateController.cs:26]: Do we need to also clear this cache if a workflow gets updated/added/removed?

Comment (skamphuis) [reply DNN Platform/Modules/HTML/Components/WorkflowStateController.cs:26]: Probably, but I haven't found any code that does such a thing...

Comment (skamphuis) [reply DNN Platform/Modules/HTML/Components/WorkflowStateController.cs:26]: @bdukes  Would it make sense to add an optional clearCache parameter to the method so extensions can control the cache themselves? Or a new method "ClearWorkflowsCache" or something?

Comment (bdukes) [reply DNN Platform/Modules/HTML/Components/WorkflowStateController.cs:26]: You're right, there isn't any code to modify these HTML workflows (in contrast to the content workflows used by the platform). Clearing the cache would be the responsibility of whoever is updating that table. I'm okay with this as-is.

**Review by bdukes — APPROVED** [review]

**Comment by mitchelsellers** [comment]
@valadas can we get another review from you to clear the requested changes review

**Review by valadas — APPROVED** [review]
Awesome

**Pull request merged**